### PR TITLE
Allow for reporting a facility as closed or reopened

### PR DESCRIPTION
**Pull request by TaiWilkin**
## Overview

Adds models and an API to support marking a facility as closed or opened.

Connects #1136 

## Demo

Screen Shot 2021-02-03 at 11 39 59 AM
Screen Shot 2021-02-03 at 11 39 36 AM
Screen Shot 2021-02-03 at 11 39 13 AM

## Testing Instructions

* Run `vagrant ssh` and `./scripts/manage migrate` and `./scripts/server`
* Go to [:6543](http://localhost:6543/) and save the OAR ID of a facility 
- Go to http://localhost:8081/api/docs/#!/facilities/facilities_read and authenticate your user.
- Fetch the facility by the ID you saved
    - The facility should have `is_closed: null`
- Go to http://localhost:8081/api/docs/#!/facilities/facilities_report
- Enter the facility id and 
    ``` 
    {
         "closure_state": "CLOSED",
         "reason_for_report": "This facility was closed."
     } 
     ```
- Submit. 
    - The request should return a FacilityActivityReport with `closure_state: ‘CLOSED’` and `status: ‘PENDING’` and appropriate user, contributor, and created/updated date times. Note the ID.
- Go to http://localhost:8081/api/docs/#!/facility-activity-reports/facility_activity_reports_approve and enter the report ID and ```{
    "status_change_reason": "The facility report was confirmed."
}```
- Submit. 
    - The request should return the same FacilityActivityReport with `status: “CONFIRMED”`, all of the status_change_ fields filled out, and `approved_at` set to the current date time.
- Go to http://localhost:8081/api/docs/#!/facilities/facilities_read and fetch the facility again
    - It should have `is_closed: true`
- Repeat the process with a new report, but enter `"closure_state": “OPEN”`
    - Approve the facility report. The facility should now have `is_closed: false`
- Create a new facility report with `"closure_state": “CLOSED”`
- Go to http://localhost:8081/api/docs/#!/facility-activity-reports/facility_activity_reports_reject 
- Enter the facility ID and ```{
    "status_change_reason": "The facility report is incorrect."
}```
    - Submit. The response should have `"status": "REJECTED”` and the status_change_ fields should be filled out appropriately. 
- Fetch the facility. 
     - Its status should not have changed. 

## Checklist

- [x] `fixup!` commits have been squashed
- [ ] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

### CODE REVIEW

**Comment by jwalgran** [comment]
Looking at this now.

File: src/django/api/views.py [hunk 1]
```diff
@@ -3263,3 +3309,110 @@ def retrieve(self, request, pk=None):
...
+
+        {
+            "status_change_reason": "CLOSED"
+        }
+    """
+    queryset = FacilityActivityReport.objects.all()
+    serializer_class = FacilityActivityReportSerializer
+    permission_classes = (IsRegisteredAndConfirmed,)
+
+    @action(detail=True, methods=['PUT'],
```
Comment (jwalgran) [inline src/django/api/views.py:3359]: For consistency with the rest of the API methods we should consider changing the path+verb combination for the approve and reject endpoints.

Pending facility matches have /approve and /reject endpoints that response to POST. It is conventional that "action" endpoints, where the path ends in a word rather than a resource ID, are POST requests.

PUT is actually a reasonable verb choice in this context, because we are updating an existing resource, not adding a new one. If we continue to use PUT, we should consider changing the path to `PUT http://localhost:8081/api/facility-activity-reports/1`.

I don't have strong feelings, but I lean toward changing the verb to POST for consistency with the facility match approve/reject endpoints.

File: src/django/api/views.py [hunk 2]
```diff
@@ -3263,3 +3309,110 @@ def retrieve(self, request, pk=None):
...
+            permission_classes=(IsRegisteredAndConfirmed,),
```
Comment (jwalgran) [inline src/django/api/views.py:3360]: We should change this permission class to `IsAdminUser` since we don't want to allow people to approve their own reports.

File: src/django/api/views.py [hunk 3]
```diff
@@ -2084,6 +2098,38 @@ def dissociate(self, request, pk=None):
...
+            facility = Facility.objects.get(pk=pk)
+        except Facility.DoesNotExist:
+            raise NotFound('Facility with OAR ID {} not found'.format(pk))
+
+        contributor = request.user.contributor
+        facility_activity_report = FacilityActivityReport.objects.create(
+            facility=facility,
+            reported_by_user=request.user,
+            reported_by_contributor=contributor,
+            closure_state=request.data.get('closure_state'),
```
Comment (jwalgran) [inline src/django/api/views.py:2126]: I was able to submit a report with `closure_state` set to "FOO". As I understand it Django's validation does not kick in unless `full_clean` is called on the model. 

Screen Shot 2021-02-04 at 10 23 21 AM

**Review by jwalgran — COMMENTED** [review]
The structure of this update is solid and I was able to successfully run through the testing instructions and exercise the workflow. There are a few changes that I think we need to make before merging.

File: src/django/api/views.py (hunk 1 above)
Comment (TaiWilkin) [reply src/django/api/views.py]: I went with `POST` for consistency. Thanks for explaining why it's used in this case!

File: src/django/api/views.py (hunk 3 above)
Comment (TaiWilkin) [reply src/django/api/views.py:2126]: I added a `full_clean` before `save()` for this call. I ended up adding `blank=True` to a few of the fields that are nullable because they were giving validations errors after adding the check, so I updated the migration.

**Comment by TaiWilkin** [comment]
Ready for another look

File: src/django/api/views.py [hunk 4]
```diff
@@ -2125,6 +2125,7 @@ def report(self, request, pk=None):
             reported_by_contributor=contributor,
             closure_state=request.data.get('closure_state'),
             reason_for_report=request.data.get('reason_for_report'))
+        facility_activity_report.full_clean()
```
Comment (jwalgran) [inline src/django/api/views.py:2128]: It looks like we need to go a bit further with this validation. When I tried POSTing "foo" as a status the API returned a full Django HTTP unhandled exception page.

Screen Shot 2021-02-08 at 11 08 00 AM

We have some existing code that handles a similar situation, but we will want to use a different exception.

https://github.com/open-apparel-registry/open-apparel-registry/blob/34ecf9d3e4a71235c4cd0fa3f8bbe71fd613e078/src/django/api/serializers.py#L68-L76

In summary, we have the catch the django.core.ValidationError that is raised by `full_clean` and raise an `api.exceptions.BadRequestException` instead. DRF will catch this and render a JSON response instead of returning HTML.

**Review by jwalgran — COMMENTED** [review]
Updates to the model and the HTTP verbs look great. There is one more adjustment to the status code validation that I would recommend to ensure that the error response is JSON and has a 400 status code.

File: src/django/api/views.py (hunk 4 above)
Comment (TaiWilkin) [reply src/django/api/views.py]: Updated this in c65d115. Thanks for the assistance!

**Comment by TaiWilkin** [comment]
Ready for another look.

**Review by jwalgran — APPROVED** [review]
Nice fixes. Sorry for the start/stop review process and thanks for working through the changes.

**Comment by TaiWilkin** [comment]
Thank you for all the help on this!
